File: catalogue-api/src/main/java/io/quartic/catalogue/api/DatasetConfig.java
package io.quartic.catalogue.api;

import com.fasterxml.jackson.annotation.JsonAnyGetter;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.quartic.common.SweetStyle;
import org.immutables.value.Value;

import java.util.Map;

@SweetStyle
@Value.Immutable
@JsonSerialize(as = DatasetConfigImpl.class)
@JsonDeserialize(as = DatasetConfigImpl.class)
@JsonIgnoreProperties(ignoreUnknown = true)
public interface DatasetConfig {
    DatasetMetadata metadata();
    DatasetLocator locator();

    @JsonAnyGetter
    Map<String, Object> extensions();
}

File: catalogue-api/src/main/java/io/quartic/catalogue/api/DatasetConfig.kt
package io.quartic.catalogue.api

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonIgnoreProperties

@JsonIgnoreProperties(ignoreUnknown = true)
data class DatasetConfig(
    val metadata: DatasetMetadata,
    val locator: DatasetLocator,
    @get:JsonAnyGetter
    val extensions: Map<String, Any>
)
